### ISSUE TITLE
Fails when mapping ports - "Failure in a Window system call, .. cannot find the file specified" when mapping ports

### ISSUE THREAD

**Issue by TheoAndersen** [docker/for-win#212]
Hi,

When i add map ports when running a container, docker crashes with a windows error (see below). I'm running Windows 10 and docker for windows beta 1.23.9-beta29.3.

Error when i add a port mapping:
```
PS C:\WINDOWS\system32> docker run -it microsoft/windowsservercore -p 8000:80
C:\Program Files\Docker\Docker\Resources\bin\docker.exe: Error response from daemon: container af38e8f3
failure in a Windows system call: The system cannot find the file specified. (0x2) extra info: {"Applic
:true,"CreateStdInPipe":true,"CreateStdOutPipe":true,"CreateStdErrPipe":false,"ConsoleSize":[42,103]}.
```

Not i did do a ```Get-ContainerNetwork | Remove-ContainerNetwork```at one time, which i got from another Issue (https://github.com/docker/compose/issues/4024) - but now the result of that seems to be that i get a windows error each time docker maps ports.

Is there any known way to rectify this?

Docker version
```
PS C:\Users\Theo Carton> docker version
Client:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Thu Oct 27 00:09:21 2016
 OS/Arch:      windows/amd64
 Experimental: true

Server:
 Version:      1.12.2-cs2-ws-beta
 API version:  1.25
 Go version:   go1.7.1
 Git commit:   050b611
 Built:        Tue Oct 11 02:35:40 2016
 OS/Arch:      windows/amd64
```

Windows version:
```
PS C:\WINDOWS\system32> [System.Environment]::OSVersion.Version

Major  Minor  Build  Revision
-----  -----  -----  --------
10     0      14393  0
```

**Comment by friism** [comment]
ping @jmesser81

**Comment by JMesser81** [comment]
It sounds like you deleted the default NAT network behind the scenes when executing the Get-ContainerNetwork | Remove-ContainerNetwork command. 

Can you restart the Docker service to see if this will re-create the NAT network? 

If you have configured your system to not create a default nat network (e.g. bridge: "none" option), then please create a new nat network using the docker network create -d nat ... command

**Comment by TheoAndersen** [comment]
Stopping the service com.docker.service (from the services MMC) resulted in this error (i think i've experienced this before):

```
"Windows could not stop the Docker for Windows Service service on Local Computer" 
Error 1067: The process terminated unexpectedly
```

But the service did stop, and i was able to start it again. Though that 1067 error came every time i tried to stop the service. 

Stopping and starting the service (not the app on the the taskbar) made docker run give this error immediately:
```
PS C:\Users\Theo Carton> docker run -it microsoft/windowsservercore -p 800:800
C:\Program Files\Docker\Docker\Resources\bin\docker.exe: An error occurred trying to connect: Post http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.24/containers/create: open //./pipe/docker_engine: The system c                                        annot find the file specified..
See 'C:\Program Files\Docker\Docker\Resources\bin\docker.exe run --help'.
```

Quitting docker from the task-bar and starting it again made it work again.
... though still with the same error when port mapping

```
PS C:\Users\Theo Carton> docker run -it microsoft/windowsservercore -p 800:800
C:\Program Files\Docker\Docker\Resources\bin\docker.exe: Error response from daemon: container 5ef507a4bad7b23d38aac2f5b75e4b69fb202fb62d95742dc0b3245728ccd777 encountered an error during CreateProcess: failure in a Windows system call: The system cannot find the file specified. (0x2) extra info: {"ApplicationName":"","CommandLine":"-p 800:800","WorkingDirectory":"C:\\","Environment":{},"EmulateConsole":true,"CreateStdInPipe":true,"CreateStdOutPipe":true,"CreateStdErrPipe":false,"ConsoleSize":[56,243]}.
```

I haven't changed the network configuration, as far as i know.
My nat configuration looks like this
```
> docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
03112863db75        nat                 nat                 local
56fd477d90a3        none                null                local

> docker network inspect nat
[
    {
        "Name": "nat",
        "Id": "03112863db751b1a3ab5b947042f16fbc1322851542703644ebd9e15668ed7b2",
        "Scope": "local",
        "Driver": "nat",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "10.0.76.0/24",
                    "Gateway": "10.0.76.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Containers": {},
        "Options": {
            "com.docker.network.windowsshim.hnsid": "1b014650-62a3-463a-b722-18b6381d8898",
            "com.docker.network.windowsshim.networkname": "nat"
        },
        "Labels": {}
    }
]
```

**Comment by JMesser81** [comment]
Which Docker Engine are you targeting? The Docker for Windows (Moby Linux VM) or local Docker Engine daemon? It sounds like you're connecting to a different engine or perhaps using the wrong connection string. (e.g. -H "localhost:2375" or -H "npipe://.pipe/win_engine). I'm not sure what the default way to connect to each docker Engine is: either TCP socket or named pipe.

**Comment by TheoAndersen** [comment]
I'm actually not sure.. All i've done is to install the "Dock for windows -beta" and right clicked the docker whale in the task-bar and selected Windows Containers. I haven't set any connection string manually.
[service.txt](https://github.com/docker/for-win/files/581800/service.txt)


Where can i figure out which connection string I'm using?

A little more info about my setup, which might help.

Also I've attached the service.txt log file (from c:/ProgramDAta/Docker/service.txt, which seems to be updated every time i run a command..)

```
docker info
Containers: 27
 Running: 0
 Paused: 0
 Stopped: 27
Images: 11
Server Version: 1.12.2-cs2-ws-beta
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: nat null overlay
Swarm: inactive
Security Options:
Kernel Version: 10.0 14393 (14393.447.amd64fre.rs1_release_inmarket.161102-0100)
Operating System: Windows 10 Enterprise
OSType: windows
Architecture: x86_64
CPUs: 6
Total Memory: 6.542 GiB
Name: DESKTOP-98GQ7ES
ID: NHHC:OE3M:TPDH:LZFI:JTT4:VOF4:5HTF:6TTK:HIXD:ZQAV:LXVP:RAGR
Docker Root Dir: C:\ProgramData\Docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
```

**Comment by friism** [comment]
@TheoAndersen @JMesser81 with Docker for Windows, shutting down the Docker for Windows app (using the whale tray icon) will have the same effect of stopping the Windows Docker daemon

**Comment by TheoAndersen** [comment]
Now on the latest beta of Docker for win, i still get an error when i run containers (now also without port mappings)

Is there anything i can do to reset my windows container/nat/hyper-v setup to default? i have a feeling this might help

```
PS C:\WINDOWS\system32> docker run -dit microsoft/windowsservercore
b203a0524fd20753b6fb19fdde3b134c10d6ed5d94abd642cb63a30c617d1703
C:\Program Files\Docker\Docker\Resources\bin\docker.exe: Error response from daemon: container b203a0524fd20753b6fb19fdd
e3b134c10d6ed5d94abd642cb63a30c617d1703 encountered an error during Start: failure in a Windows system call: The system
cannot find the path specified. (0x3): Are you trying to mount a directory onto a file (or vice-versa)? Check if the spe
cified host path exists and is the expected type.
```

**Comment by dgageot** [comment]
@TheoAndersen Can you please close Docker for Windows, run this is an elevated Powershell and then restart Docker for Windows?

```
Get-ContainerNetwork | Remove-ContainerNetwork -Force -ea SilentlyContinue
Get-NetNat | Remove-NetNat -ea SilentlyContinue
```

**Comment by TheoAndersen** [comment]
I tried the powershell commands, and exited docker, and upgraded to the latest version 1.13.0-rc2-beta31. Firstly this made me able to run containers again, though only without mapping ports.

When i map ports this happens
```
PS C:\WINDOWS\system32> Get-ContainerNetwork

Name Id                                   Subnets          Mode SourceMac DNSServers DNSSuffix
---- --                                   -------          ---- --------- ---------- ---------
nat  fb5792a6-029c-4a5c-8287-1e1ba868504c {172.29.64.0/20} NAT


PS C:\WINDOWS\system32> docker run -it microsoft/windowsservercore -p 8001:8001
C:\Program Files\Docker\Docker\Resources\bin\docker.exe: Error response from daemon: container e25fc3f3ae3ee91eb5f4c9856
18aec7f9f092e9f6e013d5cc7e19b831e166754 encountered an error during CreateProcess: failure in a Windows system call: The
 system cannot find the file specified. (0x2) extra info: {"ApplicationName":"","CommandLine":"-p 8001:8001","User":"","
WorkingDirectory":"C:\\","Environment":{},"EmulateConsole":true,"CreateStdInPipe":true,"CreateStdOutPipe":true,"CreateSt
dErrPipe":false,"ConsoleSize":[50,120]}.
PS C:\WINDOWS\system32>
```

Any suggestions would be helpful, i have no idéa what this error means. I've also attached the docker logs (cleaned out the dir prior to running the container with port mapping)

[log.txt](https://github.com/docker/for-win/files/632092/log.txt)

**Comment by TheoAndersen** [comment]
I just found the Debug-ContainerHost script and gave it a try. (https://github.com/Microsoft/Virtualization-Documentation/tree/master/windows-server-container-tools/Debug-ContainerHost)

```
PS C:\Users\Theo Carton> .\Debug-ContainerHost.ps1 > debugContainerHostLog.txt
Describing Windows Version and Prerequisites
 [+] Is Windows 10 Anniversary Update or Windows Server 2016 92ms
 [+] Has KB3192366, KB3194496, or later installed if running Windows build 14393 77ms
 [+] Is not a build with blocking issues 73ms
Describing Docker is installed
 [+] A Docker service is installed - 'Docker' or 'com.Docker.Service'  82ms
 [+] Service is running 71ms
 [+] Docker.exe is in path 2.09s
Describing User has permissions to use Docker daemon
 [+] docker.exe should not return access denied 65ms
Describing Windows container settings are correct
 [+] Do not have DisableVSmbOplock set to 1 67ms
 [+] Do not have zz values set 69ms
Describing The right container base images are installed
 [+] At least one of 'microsoft/windowsservercore' or 'microsoft/nanoserver' should be installed 152ms
Describing Container network is created
 [+] At least one local container network is available 1.16s
 [+] No more than one NAT network 83ms
 [-] NAT Network's internal prefix does not overlap with external IP' 62ms
   Expected: value was {}, but should not have been the same
   155:           $BinaryIP.Substring(0, $Length) | Should Not Be $BinaryIPSubnet.Substring(0, $Length)
   at <ScriptBlock>, C:\Users\Theo Carton\Debug-ContainerHost.ps1: line 155
   at <ScriptBlock>, C:\Users\Theo Carton\Debug-ContainerHost.ps1: line 150
```

The NAT Network internal prefix does not overlap with external IP, bit failed. Would sounds a bit like something that could be related to port mapping.

A little more docker network info
```
PS C:\WINDOWS\system32> docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
7e4e034291ce        nat                 nat                 local
53fadf940b85        none                null                local

PS C:\WINDOWS\system32> docker.exe network inspect 7e4e034291ce
[
    {
        "Name": "nat",
        "Id": "7e4e034291ce47ff1fbf68181e608c3dd994645cf7c945576b8c2fc5c70800e0",
        "Created": "2016-12-05T20:20:14.6893741+01:00",
        "Scope": "local",
        "Driver": "nat",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "windows",
            "Options": null,
            "Config": [
                {
                    "Subnet": "0.0.0.0/0",
                    "Gateway": "0.0.0.0"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Containers": {},
        "Options": {
            "com.docker.network.windowsshim.hnsid": "fb5792a6-029c-4a5c-8287-1e1ba868504c",
            "com.docker.network.windowsshim.networkname": "nat"
        },
        "Labels": {}
    }
]
```

Attached are the log output from the debugContainerHost run
[debugContainerHostLog.txt](https://github.com/docker/for-win/files/632127/debugContainerHostLog.txt)

**Comment by TheoAndersen** [comment]
I think i figured it out...

Changing the darn command, to make the port mapping part come before the image... makes it work...

So instead of 
```
docker run -it microsoft/windowsservercore -p 800:800
```
Which fails with a strange error.

I'll do
```
docker run -it -p 800:800 microsoft/windowsservercore 
```

which works.

Sorry for the trouble :/
